segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -547,366 +678,538 @@ private void validateSystemSnapshot(SystemSnapshotRecord systemSnapshot) {
+
+                            val fromOffset = new AtomicLong();
+                            val remaining = new AtomicInteger(retValue.get().length);
+
+                            // Continue until there is still data remaining to be read.
+                            return Futures.loop(
+                                        () -> remaining.get() > 0,
+                                        () -> chunkStorage.read(h, fromOffset.get(), remaining.get(), retValue.get(), Math.toIntExact(fromOffset.get())),
+                                        bytesRead -> {
+                                            remaining.addAndGet(-bytesRead);
+                                            fromOffset.addAndGet(bytesRead);
+                                        },
+                                        executor)
+                                    .thenAcceptAsync(v -> isReadDone.set(true), executor);
+
+                        }, executor)
+                        .handleAsync((v, e) -> {
+                            attempt.incrementAndGet();
+                            if (e != null) {
+                                // record the exception
+                                lastException.set(e);
+                                log.warn("SystemJournal[{}] Error while reading journal {}.", containerId, snapshotFile, lastException);
+                                return null;
+                            } else {
+                                // no exception, return the value.
+                                return v;
+                            }
+                        }, executor),
+                executor)
+                .handleAsync((v, e) -> {
+                    // If read is not done and we have exception then throw.
+                    if (!isReadDone.get() && lastException.get() != null) {
+                        throw new CompletionException(lastException.get());
+                    }
+                    return v;
+                }, executor)
+                .thenApplyAsync(v -> retValue.get(), executor);
     }
 
     /**
      * Process all systemLog entries to recreate the state of metadata storage system segments.
      */
-    private void applySystemLogOperations(MetadataTransaction txn,
-                                          SystemSnapshotRecord systemSnapshotRecord,
-                                          HashMap<String, Long> chunkStartOffsets,
-                                          HashMap<String, Long> finalTruncateOffsets,
-                                          HashMap<String, Long> finalFirstChunkStartsAtOffsets,
-                                          HashSet<SystemJournalRecord> visitedRecords) throws Exception {
-
-        long epochToStartScanning = 0;
-        int fileIndexToRecover = 1;
+    private CompletableFuture<Void> applySystemLogOperations(MetadataTransaction txn,
+                                                             SystemSnapshotRecord systemSnapshotRecord,
```

Comment:
I see you keep passing all these maps to various methods. Consider putting them into a single container object (i.e., `SystemLogOpContext`) and just pass that around. That will also help factoring out this code.

Reply:
done.